src/internal/invoker/disposable.ts

```diff
@@ -0,0 +1,47 @@
+import { HandlerContext } from "../../HandlerContext";
+import { logger } from "../util/logger";
+
+export function registerDisposable(ctx: HandlerContext): (callback: () => Promise<void>, description?: string) => void {
+    return (callback: () => Promise<void>, description: string) => {
+        if ((ctx as any).__disposables) {
+            (ctx as any).__disposables.push({ how: callback, what: description });
+        } else {
+            const disposables = [{ how: callback, what: description }];
+            (ctx as any).__disposables = disposables;
+        }
+    };
+}
+
+export function dispose(ctx: HandlerContext): () => Promise<void> {
```

Comment:
Same here, why return a function? Is it because you imagine the common use case will be `.then(dispose)`?

Reply:
That functions is being used when creating the context and gets assigned to `context.lifecycle.dispose`.